examples/tv-app/android/java/MessagesManager.cpp

```diff
@@ -0,0 +1,425 @@
+{
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+    VerifyOrReturn(env != nullptr, ChipLogError(Zcl, "Failed to GetEnvForCurrentThread for MessagesManager"));
+
+    VerifyOrReturn(mMessagesManagerObject.Init(managerObject) == CHIP_NO_ERROR,
+                   ChipLogError(Zcl, "Failed to init mMessagesManagerObject"));
+
+    jclass managerClass = env->GetObjectClass(managerObject);
+    VerifyOrReturn(managerClass != nullptr, ChipLogError(Zcl, "Failed to get MessagesManager Java class"));
+
+    mGetMessagesMethod = env->GetMethodID(managerClass, "getMessages", "()[Lcom/matter/tv/server/tvapp/Message;");
+    if (mGetMessagesMethod == nullptr)
+    {
+        ChipLogError(Zcl, "Failed to access MessagesManager 'getMessages' method");
+        env->ExceptionClear();
+    }
+
+    mPresentMessagesMethod = env->GetMethodID(
+        managerClass, "presentMessages",
+        "(Ljava/lang/String;Lcom/matter/tv/server/tvapp/Message$PriorityType;IJILjava/lang/String;Ljava/util/Vector;)Z");
+    if (mPresentMessagesMethod == nullptr)
+    {
+        ChipLogError(Zcl, "Failed to access MessagesManager 'presentMessages' method");
+        env->ExceptionClear();
+    }
+
+    mCancelMessagesMethod = env->GetMethodID(managerClass, "cancelMessage", "(Ljava/lang/String;)Z");
+    if (mCancelMessagesMethod == nullptr)
+    {
+        ChipLogError(Zcl, "Failed to access MessagesManager 'cancelMessage' method");
+        env->ExceptionClear();
+    }
+}
+
+uint32_t MessagesManager::GetFeatureMap(chip::EndpointId endpoint)
+{
+    if (endpoint >= MATTER_DM_CONTENT_LAUNCHER_CLUSTER_SERVER_ENDPOINT_COUNT)
+    {
+        return kEndpointFeatureMap;
+    }
+
+    BitMask<Feature> FeatureMap;
+    FeatureMap.Set(Feature::kReceivedConfirmation);
+    FeatureMap.Set(Feature::kConfirmationResponse);
+    FeatureMap.Set(Feature::kConfirmationReply);
+    FeatureMap.Set(Feature::kProtectedMessages);
+
+    uint32_t featureMap = FeatureMap.Raw();
+    // forcing to all features since this implementation supports all
+    // Attributes::FeatureMap::Get(endpoint, &featureMap);
+    return featureMap;
+}
+
+CHIP_ERROR MessagesManager::HandleGetMessages(AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    JNIEnv * env   = JniReferences::GetInstance().GetEnvForCurrentThread();
+    VerifyOrReturnError(env != nullptr, CHIP_JNI_ERROR_NULL_OBJECT, ChipLogError(Zcl, "Could not get JNIEnv for current thread"));
+    JniLocalReferenceScope scope(env);
+
+    ChipLogProgress(Zcl, "Received MessagesManager::HandleGetMessages");
+    VerifyOrExit(mMessagesManagerObject.HasValidObjectRef(), err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mGetMessagesMethod != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    return aEncoder.EncodeList([this, env](const auto & encoder) -> CHIP_ERROR {
+        jobjectArray messagesList = (jobjectArray) env->CallObjectMethod(mMessagesManagerObject.ObjectRef(), mGetMessagesMethod);
```

Comment:
usually we do DeviceLayer::StackUnlock unlock; before calling into java from jni in callback flow.

Reply:
applied this convention everywhere in the tv-app code